### PR TITLE
Add convTranspose2d implementation.

### LINKED ISSUE #10: Implement convTranspose2d
https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-convtranspose2d

### PR DESCRIPTION
**Pull request by BruceDai**
fix #10
@huningxin @fdwr I use **conv2d** logic to compute **convTranspose2d** with padding input of **convTranspose2d**, PTAL, thanks.

About implementation of **dilation** part, I refer to [PyTorch CONVTRANSPOSE2D](https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html), and verify the implementation by [an ONNX's  convTranspose2d test](https://github.com/onnx/onnx/blob/main/onnx/backend/test/case/node/convtranspose.py#L328).

### CODE REVIEW

File: src/conv_transpose2d.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import {transpose} from './transpose.js';
+
+/**
+ * Compute a 2-D transposed convolution given 4-D input and filter tensors.
+ * @param {Tensor} input
+ * @param {Tensor} filter
+ * @param {MLConvTranspose2dOptions} options
+ * @return {Tensor}
+ */
+export function convTranspose2d(input, filter, {padding = [0, 0, 0, 0],
```
Comment (fdwr) [inline src/conv_transpose2d.js:15]: [minor] Propose wrapping `padding` consistently to the same line as the others (strides, groups).

File: src/conv_transpose2d.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  autoPad = 'explicit',
+}
+= {}) {
+  if (inputLayout === 'nhwc') {
+    // nhwc -> nchw
+    input = transpose(input, {permutation: [0, 3, 1, 2]});
+  }
+  if (filterLayout === 'iohw') {
+    // iohw -> oihw
+    filter = transpose(filter, {permutation: [1, 0, 2, 3]});
```
Comment (fdwr) [inline src/conv_transpose2d.js:34]: Hmm, I'm confused by the filter nomenclature (granted, ConvTranspose is one of my least familiar operators) because the default is "iohw" in the spec, but the code is transposing it to "oihw", which isn't an [enum in the list](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-convtranspose2d). I'm not saying it should be changed in either this code or in the spec, just curious that the algorithm below isn't coded for the default filter layout, given that is presumably the more common one found in models. Is it more efficient to do so?

- [ONNX uses](https://github.com/onnx/onnx/blob/main/docs/Operators.md#convtranspose) filter shape order "iohw" (C x M/group x kH x kW).
- [TF uses](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) filter shape order "hwio" `[filter_height, filter_width, in_channels, out_channels]`.

File: src/conv_transpose2d.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  validateConv2dParams(input, filter, {groups, bias});
+
+  const [batchCount, inputChannels, inputHeight, inputWidth] = input.shape;
+  const [outputChannels, , filterHeight, filterWidth] = filter.shape;
+  const [strideHeight, strideWidth] = strides;
+  const [dilationHeight, dilationWidth] = dilations;
+  const effectiveFilterHeight = filterHeight + (filterHeight - 1) * (dilationHeight - 1);
+  const effectiveFilterWidth = filterWidth + (filterWidth - 1) * (dilationWidth - 1);
```
Comment (fdwr) [inline src/conv_transpose2d.js:50]: I've more commonly seen the simpler equation used:

`const effectiveFilterWidth = (filterWidth - 1) * dilationWidth + 1;`

e.g. width = 5, dilation = 3
`5 + (5 - 1) * (3 - 1) = 5 + 4 * 2 = 13`
`(5 - 1) * 3 + 1 = 4 * 3 + 1 = 13`

File: test/conv_transpose2d_test.js [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+import {convTranspose2d} from '../src/conv_transpose2d.js';
+import {clamp} from '../src/clamp.js';
+import {leakyRelu} from '../src/leaky_relu.js';
+import {relu} from '../src/relu.js';
+import {sigmoid} from '../src/sigmoid.js';
+import {Tensor} from '../src/lib/tensor.js';
+
+import * as utils from './utils.js';
+
+describe('test convTranspose2d', function() {
```
Comment (fdwr) [inline test/conv_transpose2d_test.js:12]: What are `describe` and `it`? It's a different way of listing the test data than most *previous* files, which were more data driven. (not opposing it, but curious)

File: src/conv_transpose2d.js (hunk 3 above)
Comment (BruceDai) [reply src/conv_transpose2d.js:50]: > const effectiveFilterWidth = (filterWidth - 1) * dilationWidth + 1;

Yes, it is more imaginable. 

Current WebNN API Spec described [convTranspose2d](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-conv2d): 

> output size = 1 + (input size - filter size - (filter size - 1) * (dilation - 1) + beginning padding + ending padding) / stride

Maybe we need update to
> output size = 1 + (input size - (filter size - 1) * dilation - 1 + beginning padding + ending padding) / stride

same to compute the [`output size` of `conv2d`](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-conv2d), @huningxin WDYT? Thanks.

Comment (BruceDai) [reply src/conv_transpose2d.js]: I've submitted https://github.com/webmachinelearning/webnn/pull/328 to WebML CG for getting feedbacks.

File: src/conv_transpose2d.js (hunk 1 above)
Comment (BruceDai) [reply src/conv_transpose2d.js]: Updated by new commits, thanks.

**Review by fdwr — APPROVED** [review]

File: src/conv_transpose2d.js (hunk 2 above)
Comment (fdwr) [reply src/conv_transpose2d.js:38]: Ah, your newer comment explains that you're reusing the logic for Conv2d, and hence you're switching it to "oihw" to be compatible with that. 👍

**Review by huningxin — APPROVED** [review]
